### PR TITLE
Update effective sample size computation 

### LINKED ISSUE #2810: Update effective sample size computation
...without truncating it. See [blog post](http://andrewgelman.com/2018/01/18/measuring-speed-stan-incorrectly-faster-thought-cases-due-antithetical-sampling/).
Related Stan issue:
https://github.com/stan-dev/stan/issues/2450
and PR:
https://github.com/stan-dev/stan/pull/2451

### PR DESCRIPTION
**Pull request by sharanry**
Fixes #2810
Reference: https://github.com/stan-dev/stan/blob/1770e699c2b16b738123b7b0d1894d8a23505653/src/docs/stan-reference/inference.tex#L588

### CODE REVIEW

File: pymc3/diagnostics.py [hunk 1]
```diff
@@ -225,33 +225,56 @@ def get_vhat(x):
         return Vhat
 
     def get_neff(x, Vhat):
+        def get_autocor(chains, t):
+            # Calculate lagged multiseries autocorrelation
+
+            def lagged_auto_cov(Xi, t):
```
Comment (junpenglao) [inline pymc3/diagnostics.py]: Could you please move this function out and add a test for it?

Comment (sharanry) [reply pymc3/diagnostics.py]: I have instead used the `autocorr` from `stats` in the new commit

**Comment by junpenglao** [comment]
Could you please replicate some of the experiment in http://discourse.mc-stan.org/t/n-eff-bda3-vs-stan/2608/21 to demonstrate that it is now computing the right eff_n?

**Comment by sharanry** [comment]
@junpenglao What is the preferred method to compute estimated N to show the proper working for the updated n_eff computation?

**Comment by junpenglao** [comment]
I suggest sample from the same model for a few hundred times and compute the mean of `eff_n/N` from different runs. Similar to the first figure in the above discourse post.

**Comment by ColCarroll** [comment]
Sorry if this is already known, but tests are failing for valid(?) reasons. The fastest way to recreate this locally is to check on a single failure with (for example)
```
pytest -v pymc3/tests/test_posteriors.py::TestNUTSUniform::test_neff
```
or all failures with 
```
pytest -v pymc3/tests/test_posteriors.py
```
Can be hard to remember pytest invocations, so thought I'd chime in! The failure still takes ~30 seconds for me, which is a lot...

**Comment by junpenglao** [comment]
Hi @sharanry, I sensed that you might be struggling (saw your exchange with @avehtari on [discourse.mc-stan.org](http://discourse.mc-stan.org/t/n-eff-bda3-vs-stan/2608/21)). I think it might help to redirect the discussion here. The question you have about dimensions is the dimension of the Gaussian, specifically, in the simulation 1 done by @avehtari:
> Normal distribution with diagonal covariance and increasing number of dimensions
```stan
data {
  int<lower=0> D;           // number of dimensions
}
parameters {
  vector[D] mu; 
}
model {
  mu ~ normal(0, 1);
}
```
Which translate to something like below in pymc3:
```python
D = [1,2,4,8,16,32,64,128,256,512,1024]# number of dimensions, ie shape of the gaussian
for id in D:
    with pm.Model():
        mu = pm.Normal('mu', 0, 1, shape=id)
        # Each repeat used 4 chains with total of 4000 draws saved after warm-up
        tr = pm.sample(1000, tune=1000, chains=4)
    # compute eff_n using old autocorr
    ...
    # compute eff_n using new autocorr
    ...
    # save both versions of eff_n for later comparison
    ...
```
Hope this helps ;-)

**Comment by avehtari** [comment]
> The question you have about dimensions is the dimension of the Gaussian, specifically, in the simulation 1 done by @avehtari:

Oh, I thought the question was about the tests in stan repo. Now I understand the question, and @junpenglao's answer is correct.

**Comment by sharanry** [comment]
effn
Is this satisfactory @junpenglao @avehtari ?

File: pymc3/diagnostics.py [hunk 2]
```diff
@@ -231,27 +231,22 @@ def get_neff(x, Vhat):
...
+        rho = np.ones(2 * num_chains + 1)
 
-        if t % 2:
-            t -= 1
+        # Iterate over different lags of autocorrelation
+        for t in range(1, 2 * num_chains + 2):
+            auto_corr = []
+            for m in range(num_chains):
+                auto_corr.append(autocorr(x[:, m], t))
+            print(x.shape)
```
Comment (junpenglao) [inline pymc3/diagnostics.py]: I guess you forgot to delete it.

**Comment by junpenglao** [comment]
OK, I am trying to validate your implementation with the one in PyStan, and there is a mistake somewhere for computing the `eff_n` for scaler variable. I put up a notebook here: https://gist.github.com/junpenglao/aa6ea70d55d1c93fc11e45e3bf8666bd
I think it would help if you can implement it as a function first so we can check/cross compare with Stan as much as possible.

File: pymc3/diagnostics.py [hunk 3]
```diff
@@ -231,27 +231,21 @@ def get_neff(x, Vhat):
...
+            rho[t - 1] = 1. - (W - np.mean(auto_corr)) / Vhat
 
-        neff = num_chains * num_samples / (1. + 2 * rho[1:t-1].sum())
-        return min(num_chains * num_samples, np.floor(neff))
+        tHat = 1. + 2 * rho.sum()
+        neff = num_chains * num_samples / tHat
+        return abs(neff)
 
     def generate_neff(trace_values):
         x = np.array(trace_values)
```
Comment (sharanry) [inline pymc3/diagnostics.py]: np.array() function gives different outputs for trace objects and  lists of arrays. This is probably the source of the problem. Will fix it.

**Comment by sharanry** [comment]
@junpenglao  I am unable to pin point the issue, it is working fine with other scalar inputs i generated.

Here is my test: https://gist.github.com/sharanry/4592cdb25ff3dd6f944f9435be0e49fe

**Comment by junpenglao** [comment]
No problem, I will try to have a closer look this week.

**Comment by junpenglao** [comment]
@sharanry I end up reimplementing the cython code from pystan and changed the autocorr and autocov function as well. If you dont mind I will close this and open a sperate PR.

**Comment by sharanry** [comment]
@junpenglao No problem. I learned a lot working on this! :)

**Pull request closed without merging**